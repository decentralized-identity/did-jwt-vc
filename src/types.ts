import { Signer, JWTVerified, JWTHeader, JWTOptions } from 'did-jwt'

export const JWT_ALG = 'ES256K'
export const DID_FORMAT = /^did:([a-zA-Z0-9_]+):([:[a-zA-Z0-9_.-]+)(\/[^#]*)?(#.*)?$/
export const JWT_FORMAT = /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/
export const DEFAULT_CONTEXT = 'https://www.w3.org/2018/credentials/v1'
export const DEFAULT_VC_TYPE = 'VerifiableCredential'
export const DEFAULT_VP_TYPE = 'VerifiablePresentation'
export const DEFAULT_JWT_PROOF_TYPE = 'JwtProof2020'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type JwtCredentialSubject = Record<string, any>

export interface CredentialStatus {
  id: string
  type: string
}

/**
 * A JWT payload representation of a Credential
 * @see https://www.w3.org/TR/vc-data-model/#jwt-encoding
 */
export interface JwtCredentialPayload {
  iss?: string
  sub?: string
  vc: Extensible<{
    '@context': string[] | string
    type: string[] | string
    credentialSubject: JwtCredentialSubject
    credentialStatus?: CredentialStatus
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    evidence?: any
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    termsOfUse?: any
  }>
  nbf?: number
  aud?: string | string[]
  exp?: number
  jti?: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [x: string]: any
}

/**
 * A JWT payload representation of a Presentation
 * @see https://www.w3.org/TR/vc-data-model/#jwt-encoding
 */
export interface JwtPresentationPayload {
  vp: Extensible<{
    '@context': string[] | string
    type: string[] | string
    verifiableCredential?: VerifiableCredential[] | VerifiableCredential
  }>
  iss?: string
  aud?: string | string[]
  nbf?: number
  exp?: number
  jti?: string
  nonce?: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [x: string]: any
}

export type IssuerType = Extensible<{ id: string }> | string
export type DateType = string | Date
/**
 * used as input when creating Verifiable Credentials
 */
interface FixedCredentialPayload {
  '@context': string | string[]
  id?: string
  type: string | string[]
  issuer: IssuerType
  issuanceDate: DateType
  expirationDate?: DateType
  credentialSubject: Extensible<{
    id?: string
  }>
  credentialStatus?: CredentialStatus
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  evidence?: any
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  termsOfUse?: any
}

/**
 * A more flexible representation of a {@link W3CCredential} that can be used as input to methods
 * that expect it.
 */
export type CredentialPayload = Extensible<FixedCredentialPayload>

/**
 * This is meant to reflect unambiguous types for the properties in `CredentialPayload`
 */
interface NarrowCredentialDefinitions {
  '@context': string[]
  type: string[]
  issuer: Exclude<IssuerType, string>
  issuanceDate: string
  expirationDate?: string
}

/**
 * Replaces the matching property types of T with the ones in U
 */
type Replace<T, U> = Omit<T, keyof U> & U
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Extensible<T> = T & { [x: string]: any }

/**
 * This data type represents a parsed VerifiableCredential.
 * It is meant to be an unambiguous representation of the properties of a Credential and is usually the result of a transformation method.
 *
 * `issuer` is always an object with an `id` property and potentially other app specific issuer claims
 * `issuanceDate` is an ISO DateTime string
 * `expirationDate`, is a nullable ISO DateTime string
 *
 * Any JWT specific properties are transformed to the broader W3C variant and any app specific properties are left intact
 */
export type W3CCredential = Extensible<Replace<FixedCredentialPayload, NarrowCredentialDefinitions>>

/**
 * used as input when creating Verifiable Presentations
 */
export interface FixedPresentationPayload {
  '@context': string | string[]
  type: string | string[]
  id?: string
  verifiableCredential?: VerifiableCredential[]
  holder: string
  verifier?: string | string[]
  issuanceDate?: string
  expirationDate?: string
}

/**
 * A more flexible representation of a {@link W3CPresentation} that can be used as input to methods
 * that expect it.
 */
export type PresentationPayload = Extensible<FixedPresentationPayload>

interface NarrowPresentationDefinitions {
  '@context': string[]
  type: string[]
  verifier: string[]
  verifiableCredential?: Verifiable<W3CCredential>[]
}

/**
 * This data type represents a parsed Presentation payload.
 * It is meant to be an unambiguous representation of the properties of a Presentation and is usually the result of a transformation method.
 *
 * The `verifiableCredential` array should contain parsed `Verifiable<Credential>` elements.
 * Any JWT specific properties are transformed to the broader W3C variant and any other app specific properties are left intact.
 */
export type W3CPresentation = Extensible<Replace<FixedPresentationPayload, NarrowPresentationDefinitions>>

export interface Proof {
  type?: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [x: string]: any
}

/**
 * Represents a readonly representation of a verifiable object, including the {@link Proof}
 * property that can be used to verify it.
 */
export type Verifiable<T> = Readonly<T> & { readonly proof: Proof }
export type JWT = string

/**
 * A union type for both possible representations of a Credential (JWT and W3C standard)
 *
 * @see https://www.w3.org/TR/vc-data-model/#proof-formats
 */
export type VerifiableCredential = JWT | Verifiable<W3CCredential>

/**
 * A union type for both possible representations of a Presentation (JWT and W3C standard)
 *
 * @see https://www.w3.org/TR/vc-data-model/#proof-formats
 */
export type VerifiablePresentation = JWT | Verifiable<W3CPresentation>

export type VerifiedJWT = JWTVerified

/**
 * Represents the result of a Presentation verification.
 * It includes the properties produced by `did-jwt` and a W3C compliant representation of
 * the Presentation that was just verified.
 *
 * This is usually the result of a verification method and not meant to be created by generic code.
 */
export type VerifiedPresentation = VerifiedJWT & {
  verifiablePresentation: Verifiable<W3CPresentation>
}

/**
 * Represents the result of a Credential verification.
 * It includes the properties produced by `did-jwt` and a W3C compliant representation of
 * the Credential that was just verified.
 *
 * This is usually the result of a verification method and not meant to be created by generic code.
 */
export type VerifiedCredential = VerifiedJWT & {
  verifiableCredential: Verifiable<W3CCredential>
}

/**
 * Represents a tuple of a DID-URL with a `Signer` and associated algorithm.
 */
export interface Issuer {
  did: string
  signer: Signer
  alg?: string
}

/**
 * Represents the Creation Options that can be passed to the createVerifiableCredentialJwt method.
 */
export interface CreateCredentialOptions extends Partial<JWTOptions> {
  /**
   * Determines whether the JSON->JWT transformation will remove the original fields from the input payload.
   * See https://www.w3.org/TR/vc-data-model/#jwt-encoding
   *
   * @default true
   */
  removeOriginalFields?: boolean

  /**
   * Allows including or overriding some header parameters for the resulting JWT.
   * If the issuer or holder does not list an `alg`, then the one specified in `header` will be used
   */
  header?: Partial<JWTHeader>

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [x: string]: any
}

/**
 * Represents the Verification Options that can be passed to the verifyCredential method.
 * These options are forwarded to the lower level verification code
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type VerifyCredentialOptions = Record<string, any>

/**
 * Represents the Verification Options that can be passed to the verifyPresentation method.
 * The verification will fail if given options are NOT satisfied.
 */
export interface VerifyPresentationOptions extends VerifyCredentialOptions {
  domain?: string
  challenge?: string
}

/**
 * Represents the Creation Options that can be passed to the createVerifiablePresentationJwt method.
 */
export interface CreatePresentationOptions extends CreateCredentialOptions {
  domain?: string
  challenge?: string
}
