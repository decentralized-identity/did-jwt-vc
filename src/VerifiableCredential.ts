  // tslint:disable: variable-name
import { Signer } from 'did-jwt'
import { createVerifiableCredential } from '.';
import { CredentialSubject, VerifiableCredentialPayload } from './types';

export class VerifiableCredentialBuilder {
  private _signer: Signer
  private _subject: string
  private _issuer: string
  private _credentialSubject: CredentialSubject
  private _context: string[] = ['https://www.w3.org/2018/credentials/v1']
  private _type: string[] = ['VerifiableCredential']
  private _validFrom?: number
  private _validUntil?: number
  private _expiresIn?: number
  private _id?: string

  async build(): Promise<string> {
    if (this._signer === undefined) throw new Error('signer must be set before calling build()')
    if (this._issuer === undefined) throw new Error('issuer must be set before calling build()')
    if (this._subject === undefined) throw new Error('subject must be set before calling build()')
    if (this._credentialSubject === undefined) throw new Error('credentialSubject must be set before calling build()')
    const payload: VerifiableCredentialPayload = {
      sub: this._subject,
      vc: {
        '@context': this._context,
        type: this._type,
        credentialSubject: this._credentialSubject
      }
    }
    if (this._validFrom) payload.nbf = this._validFrom
    if (this._validUntil) payload.exp = this._validUntil
    else if (payload.nbf && this._expiresIn) payload.exp = payload.nbf + this._expiresIn
    if (this._id) payload.jti = this._id
    return createVerifiableCredential(payload, {
      did: this._issuer,
      signer: this._signer
    })
  }
  setSigner(signer: Signer): VerifiableCredentialBuilder {
    this._signer = signer
    return this
  }
  setSubject(subject: string): VerifiableCredentialBuilder {
    this._subject = subject
    return this
  }
  setIssuer(issuer: string): VerifiableCredentialBuilder {
    this._issuer = issuer
    return this
  }
  setCredentialSubject(credentialSubject: CredentialSubject) {
    this._credentialSubject = credentialSubject
    return this
  }
  addContext(context: string): VerifiableCredentialBuilder {
    this._context.push(context)
    return this
  }
  addType(type: string): VerifiableCredentialBuilder {
    this._type.push(type)
    return this
  }
  setValidFrom(value: number): VerifiableCredentialBuilder {
    this._validFrom = value
    return this
  }
  setValidUntil(value: number): VerifiableCredentialBuilder {
    this._validUntil = value
    return this
  }
  expiresIn(value: number): VerifiableCredentialBuilder {
    this._expiresIn = value
    return this
  }
  setId(value: string): VerifiableCredentialBuilder {
    this._id = value
    return this
  }
  get signer() {
    return this._signer
  }
  get subject() {
    return this._subject
  }
  get issuer() {
    return this._issuer
  }
  get credentialSubject() {
    return this._credentialSubject
  }
  get context() {
    return this._context
  }
  get type() {
    return this._type
  }
  get validFrom() {
    return this._validFrom
  }
  get validUntil() {
    return this._validUntil
  }
  get id() {
    return this._id
  }
}
