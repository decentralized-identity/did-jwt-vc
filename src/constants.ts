export const JWT_ALG = 'ES256K-R'
export const DID_FORMAT = /^did:([a-zA-Z0-9_]+):([:[a-zA-Z0-9_.-]+)(\/[^#]*)?(#.*)?$/
export const JWT_FORMAT = /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/
export const DEFAULT_CONTEXT = 'https://www.w3.org/2018/credentials/v1'
export const DEFAULT_TYPE = 'VerifiableCredential'
