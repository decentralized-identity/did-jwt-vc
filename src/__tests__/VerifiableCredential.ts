import faker from 'faker'

jest.mock('../index')
import { VerifiableCredentialBuilder } from '../VerifiableCredential'
import { createVerifiableCredential } from '../index'
import { Signer } from 'did-jwt'
import { DEFAULT_CONTEXT, DEFAULT_TYPE } from '../constants'
import { CredentialSubject } from '../types'

const mockedCreateVerifiableCredential = (createVerifiableCredential as unknown) as jest.Mock<
  typeof createVerifiableCredential
>

const DID_A = 'did:ethr:0xf1232f840f3ad7d23fcdaa84d6c66dac24efb198'
const DID_B = 'did:ethr:0x435df3eda57154cf8cf7926079881f2912f54db4'
const SIGNER: Signer = async (data: string) => 'signature'
const CREDENTIAL_SUBJECT: CredentialSubject = { name: 'test' }
const now = () => Math.floor(Date.now() / 1000)

describe('VerifiableCredential', () => {
  let vc: VerifiableCredentialBuilder

  beforeEach(() => {
    jest.resetAllMocks()
    vc = new VerifiableCredentialBuilder()
  })

  describe('constructor', () => {
    it('returns something', () => {
      expect(vc).toBeTruthy()
    })
    it('sets a default context', () => {
      expect(vc.context).toEqual(['https://www.w3.org/2018/credentials/v1'])
    })
    it('sets a default type', () => {
      expect(vc.type).toEqual(['VerifiableCredential'])
    })
  })

  describe('build', () => {
    describe('when required attributes have been set', () => {
      beforeEach(() => {
        vc.setSigner(SIGNER)
          .setIssuer(DID_A)
          .setSubject(DID_B)
          .setCredentialSubject(CREDENTIAL_SUBJECT)
      })
      it('calls createVerifiableCredential with required JWT payload', () => {
        vc.build()
        expect(mockedCreateVerifiableCredential).toHaveBeenCalledWith(
          expect.objectContaining({
            sub: DID_B,
            vc: expect.objectContaining({
              '@context': [DEFAULT_CONTEXT],
              type: [DEFAULT_TYPE],
              credentialSubject: CREDENTIAL_SUBJECT
            })
          }),
          {
            did: DID_A,
            signer: SIGNER
          }
        )
      })
      it('calls createVerifiableCredential with nbf in the payload if validFrom has been set', () => {
        const timestamp = now()
        vc.setValidFrom(timestamp).build()
        expect(mockedCreateVerifiableCredential).toHaveBeenCalledWith(
          expect.objectContaining({
            nbf: timestamp
          }),
          expect.anything()
        )
      })
      it('calls createVerifiableCredential with exp in the payload if validUntil has been set', () => {
        const timestamp = now()
        vc.setValidUntil(timestamp).build()
        expect(mockedCreateVerifiableCredential).toHaveBeenCalledWith(
          expect.objectContaining({
            exp: timestamp
          }),
          expect.anything()
        )
      })
      it('calls createVerifiableCredential with jti in the payload if id has been set', () => {
        const id = faker.random.word()
        vc.setId(id).build()
        expect(mockedCreateVerifiableCredential).toHaveBeenCalledWith(
          expect.objectContaining({
            jti: id
          }),
          expect.anything()
        )
      })
      it('calculates exp using validFrom and expiresIn if validUntil has not been set', () => {
        const timestamp = now()
        const interval = 60000
        vc.setValidFrom(timestamp)
          .expiresIn(interval)
          .build()
        expect(mockedCreateVerifiableCredential).toHaveBeenCalledWith(
          expect.objectContaining({
            nbf: timestamp,
            exp: timestamp + interval
          }),
          expect.anything()
        )
      })
      it('does not set exp if expiresIn has been set but validFrom has not', () => {
        vc.expiresIn(100).build()
        const args = mockedCreateVerifiableCredential.mock.calls[0]
        expect(args[0]).not.toHaveProperty('nbf')
        expect(args[0]).not.toHaveProperty('exp')
      })
      it('sets exp to validUntil over calculating it from expiresIn', () => {
        const timestamp = now()
        const validUntil = timestamp + 2000
        vc.setValidUntil(validUntil)
          .setValidFrom(timestamp)
          .expiresIn(1000)
          .build()
        expect(mockedCreateVerifiableCredential).toHaveBeenCalledWith(
          expect.objectContaining({
            nbf: timestamp,
            exp: validUntil
          }),
          expect.anything()
        )
      })
    })
    it('rejects with an error if signer has not been set', () => {
      return expect(
        vc
          .setIssuer(DID_A)
          .setSubject(DID_B)
          .setCredentialSubject(CREDENTIAL_SUBJECT)
          .build()
      ).rejects.toThrowErrorMatchingSnapshot()
    })
    it('rejects with an error if issuer has not been set', () => {
      return expect(
        vc
          .setSigner(SIGNER)
          .setSubject(DID_B)
          .setCredentialSubject(CREDENTIAL_SUBJECT)
          .build()
      ).rejects.toThrowErrorMatchingSnapshot()
    })
    it('rejects with an error if subject has not been set', () => {
      return expect(
        vc
          .setSigner(SIGNER)
          .setIssuer(DID_A)
          .setCredentialSubject(CREDENTIAL_SUBJECT)
          .build()
      ).rejects.toThrowErrorMatchingSnapshot()
    })
    it('rejects with an error if credentialSubject has not been set', () => {
      return expect(
        vc
          .setSigner(SIGNER)
          .setIssuer(DID_A)
          .setSubject(DID_B)
          .build()
      ).rejects.toThrowErrorMatchingSnapshot()
    })
  })

  describe('setSigner', () => {
    it('sets the signer of the vc', () => {
      expect(vc.setSigner(SIGNER).signer).toEqual(SIGNER)
    })
  })

  describe('setSubject', () => {
    it('sets the subject of the vc', () => {
      expect(vc.setSubject(DID_A).subject).toEqual(DID_A)
    })
  })

  describe('setIssuer', () => {
    it('sets the issuer of the vc', () => {
      expect(vc.setIssuer(DID_A).issuer).toEqual(DID_A)
    })
  })

  describe('setCredentialSubject', () => {
    it('sets the credential subject of the vc', () => {
      const value = { [faker.random.word()]: faker.random.word() }
      expect(vc.setCredentialSubject(value).credentialSubject).toEqual(value)
    })
  })

  describe('addContext', () => {
    it('adds a context to the vc', () => {
      const value = faker.random.word()
      const n = vc.context.length
      vc.addContext(value)
      expect(vc.context.length).toEqual(n + 1)
      expect(vc.context.includes(value)).toBeTruthy()
    })
  })

  describe('addType', () => {
    it('adds a type to the vc', () => {
      const value = faker.random.word()
      const n = vc.context.length
      vc.addType(value)
      expect(vc.type.length).toEqual(n + 1)
      expect(vc.type.includes(value)).toBeTruthy()
    })
  })

  describe('setValidFrom', () => {
    it('sets the validFrom timestamp of the vc', () => {
      const value = faker.random.number()
      expect(vc.setValidFrom(value).validFrom).toEqual(value)
    })
  })

  describe('setValidUntil', () => {
    it('sets the validUntil timestamp of the vc', () => {
      const value = faker.random.number()
      expect(vc.setValidUntil(value).validUntil).toEqual(value)
    })
  })

  describe('setId', () => {
    it('it sets the id of the vc', () => {
      const value = faker.random.word()
      expect(vc.setId(value).id).toEqual(value)
    })
  })
})
