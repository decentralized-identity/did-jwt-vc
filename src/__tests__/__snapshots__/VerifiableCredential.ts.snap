// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifiableCredential build rejects with an error if credentialSubject has not been set 1`] = `"credentialSubject must be set before calling build()"`;

exports[`VerifiableCredential build rejects with an error if issuer has not been set 1`] = `"issuer must be set before calling build()"`;

exports[`VerifiableCredential build rejects with an error if signer has not been set 1`] = `"signer must be set before calling build()"`;

exports[`VerifiableCredential build rejects with an error if subject has not been set 1`] = `"subject must be set before calling build()"`;
